server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpgradeService.java

```diff
@@ -93,6 +101,42 @@ public IndexMetadata upgradeIndexMetadata(IndexMetadata indexMetadata,
         return markAsUpgraded(newMetadata);
     }
 
+    public Metadata upgradeRelationMetadata(Metadata metadata) {
+        Metadata.Builder newMetadata = Metadata.builder(metadata);
+        for (IndexMetadata indexMetadata : metadata) {
+            String indexName = indexMetadata.getIndex().getName();
+            if (IndexName.isPartitioned(indexName) == false) {
+                RelationName relationName = IndexName.decode(indexName).toRelationName();
+                DocTableInfo docTable = tableFactory.create(relationName, metadata);
+                List<String> indexUUIDs = metadata.getIndices(
```

Comment:
I think this should check if it exists by looking into metadata.schemas -> relations.  I'm aware that `tableFactory.create` will use it to create the `DocTableInfo` and this ought to only "override" the existing table with data that's the same, but I'd tend to be a bit more explicit here in that this is only supposed to create the RelationMetadata.Table structure for indices and templates where it is missing

Reply:
So, if something is upgraded in the settings we shouldn't handle it here, but add code where we upgrade the `IndexMetadata` to also update `RelationMetadata.Table#parameters`?

Reply:
Good point, I didn't consider the remaining upgrade logic. 

But I'd tend to make sure that we:

- First create the RelationMetadata if missing
- Then apply remaining upgrade logic on top of RelationMetadata (and also apply changes to the partitions if applicable)

E.g. quite a bit of `MetadataIndexUpgrader` is related to updating the `mapping`, which we can all skip if `RelationMetadata.Table` is already present.

Reply:
As discussed: let's stick to the version as it is now in the PR + a comment about it overriding existing entries to apply the upgraded settings too.

Simplifying and dropping legacy logic must be done later once we no longer need the BWC

Reply:
As discussed, added comment to explain that we use the `setTable` in both cases, as currently we update the settings on `IndexMetadata` and `IndexTemplateMetadata`, so afterwards we need to make sure that `RelationMetadata` picks up those changes.